framework/test-vectors/test-vector-enumeration.md

```diff
@@ -0,0 +1,176 @@
+
+One test vector input configuration can be constructed by selecting one possible value (or possible [representative value](test-vector-enumeration.md#selecting-a-representative-input-value)) from all relevant [input dimensions](test-vector-enumeration.md#input-dimensions).
+
+All test vector input configurations can be enumerated by constructing all unique input configurations.
+
+## Expected results
+
+An expected result is a categorical value
+and an optional error description.
+
+An expected result categorical value is one of:
+* Positive (successful test scenario)
+    * `"positive-keyring"`
+* Negative encrypt (failure on encrypt)
+    * `"negative-encrypt-keyring"`
+* Negative decrypt (failure on decrypt)
+    * `"negative-decrypt-keyring"`
+
+The error description for negative scenarios
+describes the reason for the scenario's failure.
+
+### Determining expected results
+
+In contrast to [input configuration enumeration](#enumerating-input-configurations),
+which is a constructive multiplicative process,
+the expected result of a test
+is deterministic from the input configuration.
+
+The expected result of an [input configuration](#input-configuration) can be determined
+by evaluating all relevant [evaluation rules](#expected-result-evaluation-rules) for that input configuration.
+If no evaluation rules fail, the test scenario result should be `"positive-keyring"`.
+If one evaluation rule fails, the test scenario result should be the result of that evaluation rule.
+If multiple evaluation rules fail, the test scenario should not be written. See [below](#multiple-errors-todo).
+
+#### Expected result evaluation rules
+
+An evaluation rule reads the input configuration [input configuration](#input-configuration),
+and determines whether the configuration is valid
+under the conditions the rule evaluates.
+
+For example,
+the [required encryption context CMM component](complete-vectors/required-encryption-context-cmm.md#required-encryption-context-cmm-failures-on-encrypt)
+defines the conditions for encryption to fail.
+This can be interpreted as an evaluation rule,
+and the test manifest generator should implement this
+when generating test vectors using the required encryption context CMM
+to determine when these vectors should fail.
+
+These rules can be relatively informal in the spec,
+but ideally should be written programmatically,
+so a developer can write implement these rules in test manifest generation code.
+
+Evaluation rules are a useful abstraction to determine a given test scenario's expected result. A list of independent rules is flexible and maintainable. A list of rules that maintain the same interface (reads input configuration; outputs success/fail) makes evaluating these rules programmable.
+
+### Multiple errors (TODO)
+
+An input configuration SHOULD only be deterministic of a single error.
+If a single input configuration can result in multiple errors,
+it SHOULD NOT be written.
```

Comment:
Hum, I like the idea, but what if we have 2 failure cases that BOTH really should fail.
Or another way, how do we build confidence on something that is an important secondary check?

Reply:
Plan is:
* Allow configurations with multiple independent failure reasons
* Note that the error description is just a helpful message, and may not accurate describe the error
  * This "absolves" the error message of accurately choosing one of the failure reasons
*  If we later decide to codify a concrete failure reason, we should have a new attribute separate from the error description.